Add TailwindClass PropVal

oshi97 opened the pull request:
This PR adds a new custom PropVal, TailwindClass, which is used to designate special tailwind specific prop behavior, and corresponding handling in the UI.
This allows users to pick tailwind classes from their safelisted custom classes and append or remove them from a prop.

![image](https://github.com/yext/studio/assets/23005393/8ea63883-cdc3-4986-9baf-568e59b09f17)

Also updated acceptance tests to always update snapshots.
This should make life easier in the majority of cases.
If a snapshot is out of date, it will need to be updated anyway. Either way, CI will need to be manually re-run after updating the snapshots, whether that's done manually or by CI. Also for windows snapshots we rely on CI to update them since most of us don't own (or don't want to switch onto) a windows machine for updating snapshots.

J=SLAP-2876
TEST=manual,auto

can add, remove, and save TailwindClass props to file
added unit tests for new component
added new acceptance test

nmanu1 commented on packages/studio-plugin/src/utils/TypeGuards.ts:
```diff
@@ -122,6 +123,8 @@ export default class TypeGuards {
           )
         );
+      case PropValueType.TailwindClass:
+        return typeof value === "string" || value === undefined;
```
I don't think we'd always want to allow `undefined` as a valid value. If the prop is required, then the value would need to be a string. IIRC, so far we only use this typeguard for defined values. But, if we need to handle a case where the value is `undefined`, then we'd want to use `propMetadata.required` to determine if that's valid

oshi97 replied:
updated! I thought I needed this for some weird undefined button behavior but it seems to be working without it now

nmanu1 commented on e2e-tests/__screenshots__/darwin/nested-props.spec.ts/renders-nested-props-1.png:
can we show an example of a Tailwind class prop here?

oshi97 commented on packages/studio-plugin/src/utils/TypeGuards.ts:
```diff
@@ -122,6 +123,8 @@ export default class TypeGuards {
               : !metadata.required
           )
         );
+      case PropValueType.TailwindClass:
```
Not sure what the TailwindCSS class is, but any string is considered valid tailwind right now so nope

tmeyer2115 replied:
The `TailwindCSS` type is defined on line 89 of the `PropValues.ts`. If we're not using `TailwindCSS` anywhere an any string is considered valid, should we simply have: 
```
type TailwindClassProp = {
  valueType: PropValueType.TailwindClass;
  value: string;
};
```

oshi97 replied:
Oh you meant `TailwindClass`. I think it's redundant to add it since it's just the string. We also don't have HexColor in the union

oshi97 replied:
`TailwindClass` is imported by the frontend and needed to designate a prop for special tailwind behavior

oshi97 commented on packages/studio-plugin/src/utils/TypeGuards.ts:
```diff
@@ -122,6 +123,8 @@ export default class TypeGuards {
           )
         );
+      case PropValueType.TailwindClass:
+        return typeof value === "string" || value === undefined;
```
It's not possible right now, studio-plugin isn't able to read the tailwind config. I tried to but we're not able to dynamic import .ts files due vite running in cjs. I had a hack previously by using `vm.runInNewContext` which is essentially a safer eval but I don't want to do that.

From a higher level I think additional validation here would be messy, since it wouldn't be in sync with what typescript expects (which is any string) so a developer might add other tailwind classes outside of their custom theme, and we wouldn't want to error out in those cases.

oshi97 commented on packages/studio/src/components/TailwindPropInput.tsx:
```diff
@@ -0,0 +1,181 @@
+  value,
+  disabled,
+}: Props) {
+  const safelist = useAwaitedSafelist(value);
```
Where is it mutating safelist? Do you mean the .filter? I think only immutable methods are being used.
The point of the filter is to prevent the user from adding duplicates, so a `canAddClasses` or similar boolean wouldn't work

oshi97 replied:
Yes updated!

oshi97 commented on packages/studio-plugin/src/types/PropShape.ts:
```diff
@@ -35,10 +36,17 @@ type NonUnionPropType<T> = {
+type TailwindClassPropType = {
+  type: PropValueType.TailwindClass;
+  unionValues?: never;
+  customClasses?: string[];
```
oops that's from an old implementation that didn't work, updated!

oshi97 commented on packages/studio/src/components/TailwindPropInput.tsx:
```diff
@@ -0,0 +1,181 @@
+    setIsOpen(false);
+  });
+
+  const toggleOpen = useCallback(() => setIsOpen(!isOpen), [isOpen]);
```
updated!

oshi97 commented on packages/studio-plugin/src/types/PropValues.ts:
```diff
@@ -23,6 +23,7 @@ export type LiteralProp<T = PropValues> = {
+  | TailwindClassProp
 );
 
 export type ExpressionProp =
```
updated!

oshi97 commented on packages/studio/src/components/TailwindPropInput.tsx:
```diff
@@ -0,0 +1,181 @@
+).then((module) => {
+  const theme: StudioTailwindTheme | undefined = module.default?.theme?.extend;
+  return theme && generateTailwindSafelist(theme);
+});
```
added!

nmanu1 commented on packages/studio/src/components/TailwindPropInput.tsx:
```diff
@@ -0,0 +1,189 @@
+  disabled: boolean;
+}
+
+export default function TailwindPropInput({
```
do we also want to support typing into the input so people can specify non-custom Tailwind classes?

oshi97 replied:
We don't support non-custom tailwind classes right now so not yet. Right now somebody may type a tailwind class that isn't included in the current tailwind bundle on the page

nmanu1 reviewed (commented):
> ![image](https://github.com/yext/studio/assets/23005393/a008e1dc-ee6f-426c-889a-312c00fc6a8a)

I didn't check out your branch, but based on this screenshot from the PR description, it looks like the "open" state of the dropdown displays pill-like  options for custom classes below the input. It's not very clear to me that the toggle behavior would be tied to the embed icon. Does the icon gave any visual indicator when it's open vs. closed?

Also, super nit, but it doesn't look like the embed icon is exactly centered with the pills in the input?

oshi97 commented:
> > ![image](https://user-images.githubusercontent.com/23005393/262702400-a008e1dc-ee6f-426c-889a-312c00fc6a8a.png)
> 
> I didn't check out your branch, but based on this screenshot from the PR description, it looks like the "open" state of the dropdown displays pill-like options for custom classes below the input. It's not very clear to me that the toggle behavior would be tied to the embed icon. Does the icon gave any visual indicator when it's open vs. closed?
> 
> Also, super nit, but it doesn't look like the embed icon is exactly centered with the pills in the input?

updated styling! should be reflected in new snapshots assuming the manual workflow I added works

nmanu1 commented on e2e-tests/__screenshots__/darwin/custom-tailwind.spec.ts/TailwindClass-prop-editor-1.png:
Is there a reason this input is smaller than usual and the embed icon is to the left instead of the right? The same screenshot on Windows looks pretty similar to our other prop inputs

oshi97 commented on e2e-tests/__screenshots__/darwin/custom-tailwind.spec.ts/TailwindClass-prop-editor-2.png:
tweaked some styling and think things look better now. The clipping can still occur at page widths less than ~660px, but the whole page looks funky in that case so I'm not super worried about it.
I think we might be able to fix this general issue we've been seeing with clipping by wrapping the right sidebar in position: relative and then a child div with position: absolute, though I haven't tested enough to say for sure

oshi97 replied:
Also updated the styling for the undefined menu button because it felt weird how it would float off to the right side of the screen and require scrolling to get to

alextaing commented on packages/studio/src/components/ComponentNode.tsx:
```diff
@@ -79,7 +79,7 @@ export default function ComponentNode(props: ComponentNodeProps): JSX.Element {
     <div className={componentNodeClasses} style={componentNodeStyle}>
       <div
-        className="flex grow items-center cursor-pointer"
+        className="flex items-center cursor-pointer"
```
Do we need to remove this `grow`?  I without it, we will need to click the name of the component directly, instead of anywhere in that selection "bar". Will send a loom shortly.

nmanu1 commented on packages/studio/src/components/PropEditor.tsx:
```diff
@@ -41,9 +41,9 @@ export default function PropEditor({
-      <label>
+    <div className="flex items-center mb-2 text-sm grow">
+      {isNested && renderBranchUI("pb-10")}
+      <label className="flex flex-col relative grow">
```
I think some of this styling is causing the prop inputs to be clickable beyond the visible input I'd expect. Also, the undefined menu placing is a little off

https://github.com/yext/studio/assets/88398086/a4b9c258-a2c6-424c-890a-ba6405b4efc2

nmanu1 replied:
Looks like clicking on the prop name still counts as clicking on the input, which feels a little weird for the color picker. But I think it's ok if it's difficult/messier to change it back.

The undefined menu placing still needs to be tweaked though
<img width="297" alt="Screen Shot 2023-08-28 at 1 50 29 PM" src="https://github.com/yext/studio/assets/88398086/83e0bf47-8333-41c4-a188-6246252a3b7c">

nmanu1 replied:
Oh you're right, I hadn't noticed that before! It's just the undefined menu then that needs to be adjusted

oshi97 commented on packages/studio/src/components/ComponentNode.tsx:
```diff
@@ -79,7 +79,7 @@ export default function ComponentNode(props: ComponentNodeProps): JSX.Element {
   return (
     <div className={componentNodeClasses} style={componentNodeStyle}>
       <div
-        className="flex grow items-center cursor-pointer"
```
will add back!

oshi97 commented on .github/setup-acceptance/action.yml:
```diff
@@ -0,0 +1,24 @@
+    - run: |
+        npm run build -w=packages/studio-plugin
+        npm run build -w=packages/studio
+      shell: bash
```
yup github action syntax is different than the workflow one
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

oshi97 commented on .github/workflows/playwright.yml:
```diff
@@ -18,35 +18,23 @@ jobs:
-      - run: npm run playwright -- --shard=${{ matrix.shard }}/4
+      - uses: ./.github/setup-acceptance
+
+      - run: npm run update-snapshots -- --shard=${{ matrix.shard }}/4
```
no it won't, but it will put up another commit. I'll update it to fail though!

oshi97 commented on e2e-tests/__screenshots__/darwin/nested-props.spec.ts/renders-nested-props-1.png:
they're not anymore. I think this is still preferable overall

oshi97 replied:
On the other hand I think I can tweak it a bit so it lines up more

oshi97 commented on e2e-tests/tests/custom-tailwind.spec.ts:
```diff
@@ -20,3 +21,29 @@ studioTest(
+  const hasExpectedPageContents = fs
+    .readFileSync(pagePath, "utf-8")
+    .includes(expectedPageContents);
+  expect(hasExpectedPageContents).toBeTruthy();
```
I think the current way is simpler, I don't care about matching the entire file

nmanu1 commented on e2e-tests/tests/undefined-menu-button.spec.ts:
```diff
@@ -7,5 +7,4 @@ studioTest("renders undefined menu button", async ({ page, studioPage }) => {
   await expect(page).toHaveScreenshot();
   await page.getByText("Reset to Default").hover();
   await page.mouse.wheel(200, 0);
-  await expect(page).toHaveScreenshot();
```
nit: I think we can remove lines 8 and 9 too

nmanu1 commented on e2e-tests/__screenshots__/win32/nested-props.spec.ts/renders-nested-props-1.png:
does the second screenshot for this test need to be deleted?

oshi97 replied:
updated!

nmanu1 commented on e2e-tests/__screenshots__/darwin/undefined-menu-button.spec.ts/renders-undefined-menu-button-1.png:
I guess the diff is too small here for the workflow to update the snapshots automatically

tmeyer2115 commented:
For now, let's just re-generate the Windows screenshots so they have the Prominent Image.
 
Until we figure out a sane way to wait for image loading, I think we need to remove images from our screenshots. The Windows screenshots are also missing the small logo image in the Header. Having images in our screenshots, without a proper wait, is just causing spurious failures. That's definitely outside the scope of this PR, though. I'll make a separate item.

nmanu1 reviewed (approved):
Since we'll be removing the images from the screenshots anyways, lgtm!
